test/extended/baremetal/machine.go

```diff
@@ -0,0 +1,134 @@
+
+	restMapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+	scaleKindResolver := scale.NewDiscoveryScaleKindResolver(discoveryClient)
+
+	scaleClient, err := scale.NewForConfig(cfg, restMapper, dynamic.LegacyAPIPathResolverFunc, scaleKindResolver)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	return scaleClient, nil
+}
+
+func getObjects(from *objx.Value) []objx.Map {
+	var values []objx.Map
+	switch {
+	case from.IsObjxMapSlice():
+		return from.ObjxMapSlice()
+	case from.IsInterSlice():
+		for _, i := range from.InterSlice() {
+			if msi, ok := i.(map[string]interface{}); ok {
+				values = append(values, msi)
+			}
+		}
+	}
+	return values
+}
+
+func getMachineSetReplicaNumber(item objx.Map) int {
+	replicas, _ := strconv.Atoi(item.Get("spec.replicas").String())
+	return replicas
+}
+
+// getMachineName returns the machine name from given the item
+func getMachineName(item objx.Map) string {
+	return item.Get("metadata.name").String()
+}
+
+// getWorkerMachineSet returns the machineset will be used for scaling.
+func getWorkerMachineSet(dc dynamic.Interface) objx.Map {
+	mc := machineSetClient(dc)
+	obj, err := mc.List(context.TODO(), metav1.ListOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	var machineSets []objx.Map
+	for _, ms := range getObjects(objx.Map(obj.UnstructuredContent()).Get("items")) {
+		labels := (*ms.Get("spec.template.metadata.labels")).Data().(map[string]interface{})
+		if val, ok := labels["machine.openshift.io/cluster-api-machine-role"]; ok {
+			if val == "worker" {
+				machineSets = append(machineSets, ms)
+				continue
+			}
+		}
+	}
+
+	if len(machineSets) == 0 {
+		e2eskipper.Skipf("no machineset found to be used for scaling.")
+	}
+
+	// In baremetal, we'd expect only one machineset.
+	return machineSets[0]
+}
+
+func getWorkerMachineSetByName(dc dynamic.Interface, name string) objx.Map {
+	mc := machineSetClient(dc)
+	obj, err := mc.Get(context.TODO(), name, metav1.GetOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	return obj.UnstructuredContent()
+}
+
+// scaleMachineSet scales a machineSet with a given name to the given number of replicas
+func scaleMachineSet(name string, replicas int) {
+	scaleClient, err := getScaleClient()
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	s, err := scaleClient.Scales(machineAPINamespace).Get(context.TODO(), schema.GroupResource{Group: "machine.openshift.io", Resource: "MachineSet"}, name, metav1.GetOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	scaleUpdate := s.DeepCopy()
+	scaleUpdate.Spec.Replicas = int32(replicas)
+	_, err = scaleClient.Scales(machineAPINamespace).Update(context.TODO(), schema.GroupResource{Group: "machine.openshift.io", Resource: "MachineSet"}, scaleUpdate, metav1.UpdateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+// ScaleUpToExtraWorkers scales machine resources to provision
+// extra workers up to the given scale number.
+func ScaleUpToExtraWorkers(client dynamic.Interface, scale int) string {
```

Comment:
The scale up/down are interesting methods, but I'm wondering if such amount of work may be an indication that the current approach couldn't be different.
My main concern is that it seems it could be a good amount of logical overlap with the default scale&grow serial test, and obviously this functionality (scale up/down) is not a baremetal-only trait, so it could make sense to remain general.
So, given that, could make sense instead not to add a new e2e test, but a new job configuration where `virtualMediaViaExternalNetwork` is set (and run the usual serial suite)?

Reply:
All these scale up/down methods come here due to circular dependency between scale test and baremetal package. And I'm also not comfortable adding these just for this test. There are 2 options actually;
* After updating `virtualMediaViaExternalNetwork` to true, it already triggers all hosts re-provisioned and maybe there is no need to scale up extra worker. We can just check the current ones' statuses after update and revert back if it passes.
* Or like you said we can define this as a new job instead new test. That time, we'll only test `virtualMediaViaExternalNetwork` is true working or not. But not updating within running cluster.

I'm fine with both options.

Reply:
In this case, I'd opt for the second one. But I'd like to also to hear the opinions of @bfournie and @elfosardo on this topic

Reply:
I assume we need to move the host to *provisioned* to get any useful results, so we need to do the full scaling? (For example with the hostFirmwareSettings test I had started out doing a full scale up of the host but realized I could do a test just with the host in *Available*, but I I don't think that's the case here). If so I think its fine to duplicate the scale here (that's the same approach I originally took) but also the second option with a new job would probably be cleaner as suggested.

Reply:
I would also go for the second option, it should be enough in this case